### DISCUSSION TITLE
useContractRead hook return undefined value

### DISCUSSION THREAD

**Discussion by Xeway** [wevm/wagmi#376] [Q&A] [answered]
Hey,
Here's my code snippet:
```
const { DAIDecimals, isError, isLoading } = useContractRead(
        {
            addressOrName: DAIAddress,
            contractInterface: erc20ABI
        },
        "decimals"
);
```
Basically when I `console.log(DAIDecimals)` it returns me `undefined`.
DAIAddress is imported from another file, the value is good (0x4F96Fe3b7A6Cf9725f59d353F723c1bDb64CA6Aa).
The abi is imported from the library.
I'm on Kovan.

Tahnk you in advance,

**Comment by Xeway** [comment]
And also (to not create another discussions) : I get an error when for example I use:
```
const connect = useConnect({
    onConnect(data) {
      console.log('Connect', data)
}
```
Saying that useConnect takes no argument.

**Reply by SaadNiaziDev** [reply]
Looks like you missed some part of docs. OnConnect callbak is part of useAccount hook. You have to use onSuccess callback here.

**Comment by Xeway** [comment]
Ok so I used the useProvider hook to see on which network I am and wrong network.
So I changed a thing for my createClient setup:
```
const provider = ({ chainId }: WagmiProps) =>
  new providers.InfuraProvider(chainId, infuraId);

function MyApp({ Component, pageProps }: AppProps) {
  return (
  <Provider autoConnect connectors={connectors} provider={42}>
    <Component {...pageProps} />
  </Provider>
  );
}
```
I passed 42 because I want to be on Kovan. But now I get: `Error: invalid signer or provider (argument="signerOrProvider", value=42, code=INVALID_ARGUMENT, version=contracts/5.6.0)`
I'm a bit lost...
And also I don't understand why passing `connectors` as an argument in `connectors` ([my code comes from this example](https://github.com/zkSoju/wagmi-boiler/blob/main/pages/_app.tsx), and the argument taken should be a number).
I've tried YOUR example (from the doc), but I also get errors...

**Comment by Xeway** [comment]
Also, basically I've installed Redux and have data that indicate if the user is connected, on which network he is etc...
I've two pages where on 1 of them I use useNetwork. And when the user change network from his wallet's UI, it changes my Redux data.
My question is, in my second page, do I also have to code the same thing (using useNetwork), and detect when the user change network ?
At a big scale, it will be painful, idk if you have solution, thanks.

**Comment by Xeway** [comment]
Up !

**Reply by ummartoor** [reply]
did you find any solution for this error ?

**Reply by Xeway** [reply]
No, sorry :( Good luck!

**Reply by ummartoor** [reply]
my contract was on sepolia i hadn't used that in chains in rainbowkit

**Comment by ummartoor** [comment, marked as answer]
Solved
here's my code for next.js
page.js
```javascript
"use client";
import styles from "./page.module.css";
import { ConnectButton } from "@rainbow-me/rainbowkit";
import conAbi from "./conAbi.json";
import React from "react";
import { useContractRead } from "wagmi";

export default function Home() {
  const {
    data,
    isError,
    isLoading,
    isSuccess,
    refetch: refetchProjects,
  } = useContractRead({
    address: "0x710dB32Ef0E75e4868318714356773C0b28DF475",
    abi: conAbi,
    functionName: "decimals",
    onSuccess(data) {
      console.log(data);
    },
    onerror(error) {
      console.log(error);
    },
  });
  React.useEffect(() => {
    refetchProjects?.();
  }, [isSuccess]);
  return (
    <main className={styles.main}>
      <ConnectButton />
    </main>
  );
}
``` 
layout.js
/import "./globals.css";
import "@rainbow-me/rainbowkit/styles.css";
import { getDefaultWallets, RainbowKitProvider } from "@rainbow-me/rainbowkit";
import { configureChains, createConfig, WagmiConfig } from "wagmi";
import { mainnet, polygon, sepolia } from "wagmi/chains";
import { alchemyProvider } from "wagmi/providers/alchemy";
import { publicProvider } from "wagmi/providers/public";

import { Inter } from "next/font/google";

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

const { chains, publicClient } = configureChains(
  [mainnet, polygon, sepolia],
  [
    alchemyProvider({ apiKey: "AlchemyProvider_apiKey" }),
    publicProvider(),
  ]
);

const { connectors } = getDefaultWallets({
  appName: "your_app_name",
  projectId: "your_project_id",
  chains,
});

const wagmiConfig = createConfig({
  autoConnect: true,
  connectors,
  publicClient,
});

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <WagmiConfig config={wagmiConfig}>
          <RainbowKitProvider chains={chains}>{children}</RainbowKitProvider>
        </WagmiConfig>
      </body>
    </html>
  );
}

**Reply by Xeway** [reply]
Well I trust you lol and mark your post as answer. Thanks for your help!
